File: monitor/stores/walmart_ca/product.ts
import { Result } from "better-result";
import * as v from "valibot";

import { pyRun } from "@/lib/py";

const outputSchema = v.object({
  title: v.string(),
  inStock: v.boolean(),
  price: v.optional(v.number()),
  image: v.optional(v.string()),
});

type Output = v.InferOutput<typeof outputSchema>;

const config: MonitorConfig = {
  store: "walmart_ca",
  type: "product",
  description: "Monitor a Walmart Canada product by SKU",
  cron: "*/1 * * * *",
  input: {
    sku: {
      description: "Product SKU",
      type: "string",
      required: true,
      unique: true,
    },
    maxPrice: {
      description: "Maximum price to notify for",
      type: "number",
      required: true,
    },
  },
};

const run: MonitorRunner = (prev: Output, input) =>
  Result.gen(async function* () {
    const data = yield* Result.await(
      pyRun<Output>({
        store: config.store,
        type: config.type,
        sku: input.sku,
      }),
    );

    const backInStock = !prev.inStock && !!data.inStock;
    const goodPrice = (data.price ?? Infinity) <= input.maxPrice;

    const notify = backInStock && goodPrice;

    return Result.ok({
      notify,
      data,
    });
  });

// const createJobs: MonitorJobCreator = (rows) => rows.map(({ id }) => ({ id, run }));

export default {
  config,
  run,
} satisfies Monitor;

File: monitor/stores/walmart_ca/product.py
from monitor.utils.ssr import (
  extract_next_ssr_data_zendriver,
)
import zendriver as zd
import asyncio


async def main(input):
  url = f"https://www.walmart.ca/en/ip/{input['sku']}"

  browser = await zd.start(
    headless=False,
  )
  page = await browser.get(url)

  data = await extract_next_ssr_data_zendriver(page)
  item = (
    data.get("props").get("pageProps").get("initialData").get("data").get("product")
  )

  sold_by_wm = item.get("sellerId") == "0"
  in_stock = item.get("availabilityStatusV2", {}).get("value") == "IN_STOCK"

  data = {
    "title": item.get("name"),
    "inStock": sold_by_wm and in_stock,
    "price": item.get("priceInfo", {}).get("currentPrice", {}).get("price"),
    "image": item.get("imageInfo", {}).get("thumbnailUrl"),
  }
  return data


if __name__ == "__main__":
  asyncio.run(main({"sku": "70PH2R4R8A83"}))
